File: addons/guide-cpp/src/guide_action.cpp
#include "guide_action.h"
#include <godot_cpp/classes/input.hpp>
#include <godot_cpp/classes/input_event_action.hpp>
#include <godot_cpp/core/math.hpp>
#include <cmath>

using namespace godot;

void GUIDEAction::_bind_methods() {
    BIND_ENUM_CONSTANT(BOOL);
    BIND_ENUM_CONSTANT(AXIS_1D);
    BIND_ENUM_CONSTANT(AXIS_2D);
    BIND_ENUM_CONSTANT(AXIS_3D);

    BIND_ENUM_CONSTANT(TRIGGERED);
    BIND_ENUM_CONSTANT(ONGOING);
    BIND_ENUM_CONSTANT(COMPLETED);

    ClassDB::bind_method(D_METHOD("get_action_name"), &GUIDEAction::get_action_name);
    ClassDB::bind_method(D_METHOD("set_action_name", "name"), &GUIDEAction::set_action_name);
    ADD_PROPERTY(PropertyInfo(Variant::STRING_NAME, "name"), "set_action_name", "get_action_name");

    ClassDB::bind_method(D_METHOD("get_action_value_type"), &GUIDEAction::get_action_value_type);
    ClassDB::bind_method(D_METHOD("set_action_value_type", "type"), &GUIDEAction::set_action_value_type);
    ADD_PROPERTY(PropertyInfo(Variant::INT, "action_value_type", PROPERTY_HINT_ENUM, "BOOL,AXIS_1D,AXIS_2D,AXIS_3D"), "set_action_value_type", "get_action_value_type");

    ClassDB::bind_method(D_METHOD("get_block_lower_priority_actions"), &GUIDEAction::get_block_lower_priority_actions);
    ClassDB::bind_method(D_METHOD("set_block_lower_priority_actions", "block"), &GUIDEAction::set_block_lower_priority_actions);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "block_lower_priority_actions"), "set_block_lower_priority_actions", "get_block_lower_priority_actions");

    ClassDB::bind_method(D_METHOD("get_emit_as_godot_actions"), &GUIDEAction::get_emit_as_godot_actions);
    ClassDB::bind_method(D_METHOD("set_emit_as_godot_actions", "emit"), &GUIDEAction::set_emit_as_godot_actions);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "emit_as_godot_actions"), "set_emit_as_godot_actions", "get_emit_as_godot_actions");

    ClassDB::bind_method(D_METHOD("get_is_remappable"), &GUIDEAction::get_is_remappable);
    ClassDB::bind_method(D_METHOD("set_is_remappable", "remappable"), &GUIDEAction::set_is_remappable);
    ADD_PROPERTY(PropertyInfo(Variant::BOOL, "is_remappable"), "set_is_remappable", "get_is_remappable");

    ClassDB::bind_method(D_METHOD("get_display_name"), &GUIDEAction::get_display_name);
    ClassDB::bind_method(D_METHOD("set_display_name", "display_name"), &GUIDEAction::set_display_name);
    ADD_PROPERTY(PropertyInfo(Variant::STRING, "display_name"), "set_display_name", "get_display_name");

    ClassDB::bind_method(D_METHOD("get_display_category"), &GUIDEAction::get_display_category);
    ClassDB::bind_method(D_METHOD("set_display_category", "display_category"), &GUIDEAction::set_display_category);
    ADD_PROPERTY(PropertyInfo(Variant::STRING, "display_category"), "set_display_category", "get_display_category");

    ADD_SIGNAL(MethodInfo("triggered"));
    ADD_SIGNAL(MethodInfo("started"));
    ADD_SIGNAL(MethodInfo("ongoing"));
    ADD_SIGNAL(MethodInfo("completed"));
    ADD_SIGNAL(MethodInfo("cancelled"));

    ClassDB::bind_method(D_METHOD("is_triggered"), &GUIDEAction::is_triggered);
    ClassDB::bind_method(D_METHOD("is_completed"), &GUIDEAction::is_completed);
    ClassDB::bind_method(D_METHOD("is_ongoing"), &GUIDEAction::is_ongoing);
    ClassDB::bind_method(D_METHOD("get_value_bool"), &GUIDEAction::get_value_bool);
    ClassDB::bind_method(D_METHOD("get_value_axis_1d"), &GUIDEAction::get_value_axis_1d);
    ClassDB::bind_method(D_METHOD("get_value_axis_2d"), &GUIDEAction::get_value_axis_2d);
    ClassDB::bind_method(D_METHOD("get_value_axis_3d"), &GUIDEAction::get_value_axis_3d);
}

GUIDEAction::GUIDEAction() {
}

GUIDEAction::~GUIDEAction() {
}

void GUIDEAction::_triggered(Vector3 value, double delta) {
    _triggered_seconds += delta;
    _elapsed_ratio = 1.0;
    _update_value(value);
    _last_state = TRIGGERED;
    emit_signal("triggered");
    _emit_godot_action_maybe(true);
}

void GUIDEAction::_started(Vector3 value) {
    _elapsed_ratio = 0.0;
    _update_value(value);
    _last_state = ONGOING;
    emit_signal("started");
    emit_signal("ongoing");
}

void GUIDEAction::_ongoing(Vector3 value, double delta) {
    _elapsed_seconds += delta;
    if (_trigger_hold_threshold > 0) {
        _elapsed_ratio = _elapsed_seconds / _trigger_hold_threshold;
    }
    _update_value(value);
    bool was_triggered = _last_state == TRIGGERED;
    _last_state = ONGOING;
    emit_signal("ongoing");
    if (was_triggered) {
        _emit_godot_action_maybe(false);
    }
}

void GUIDEAction::_cancelled(Vector3 value) {
    _elapsed_seconds = 0;
    _elapsed_ratio = 0;
    _update_value(value);
    _last_state = COMPLETED;
    emit_signal("cancelled");
    emit_signal("completed");
}

void GUIDEAction::_completed(Vector3 value) {
    _elapsed_seconds = 0;
    _elapsed_ratio = 0;
    _triggered_seconds = 0;
    _update_value(value);
    _last_state = COMPLETED;
    emit_signal("completed");
    _emit_godot_action_maybe(false);
}

void GUIDEAction::_emit_godot_action_maybe(bool pressed) {
    if (!emit_as_godot_actions) {
        return;
    }
    if (name.is_empty()) {
        return;
    }
    Ref<InputEventAction> godot_action;
    godot_action.instantiate();
    godot_action->set_action(name);
    godot_action->set_strength(_value.x);
    godot_action->set_pressed(pressed);
    Input::get_singleton()->parse_input_event(godot_action);
}

void GUIDEAction::_update_value(Vector3 value) {
    switch (action_value_type) {
        case BOOL:
        case AXIS_1D:
            _value_bool = Math::abs(value.x) > 0;
            _value_axis_2d = Vector2(Math::abs(value.x), 0);
            _value = Vector3(value.x, 0, 0);
            break;
        case AXIS_2D:
            _value_bool = Math::abs(value.x) > 0;
            _value_axis_2d = Vector2(value.x, value.y);
            _value = Vector3(value.x, value.y, 0);
            break;
        case AXIS_3D:
            _value_bool = Math::abs(value.x) > 0;
            _value_axis_2d = Vector2(value.x, value.y);
            _value = value;
            break;
    }
}

bool GUIDEAction::is_triggered() const {
    return _last_state == TRIGGERED;
}

bool GUIDEAction::is_completed() const {
    return _last_state == COMPLETED;
}

bool GUIDEAction::is_ongoing() const {
    return _last_state == ONGOING;
}

String GUIDEAction::_editor_name() const {
    if (!display_name.is_empty()) {
        return display_name;
    }
    if (!name.is_empty()) {
        return name;
    }
    return get_path().get_file().replace(".tres", "");
}

File: addons/GuideCS/GUIDEAction.cs
using System;
using Godot;

namespace Guide;

[Tool]
[GlobalClass, Icon("res://addons/GuideCS/GUIDEAction.svg")]
public partial class GUIDEAction : Resource {
	public enum GUIDEActionValueType {
		Bool = 0,
		Axis1D = 1,
		Axis2D = 2,
		Axis3D = 3
	}

	public enum GUIDEActionState {
		Triggered,
		Ongoing,
		Completed
	}

	private StringName _name;
	[Export]
	public StringName Name {
		get => _name;
		set {
			if (_name == value) return;
			_name = value;
			EmitChanged();
		}
	}

	private GUIDEActionValueType _actionValueType = GUIDEActionValueType.Bool;
	[Export]
	public GUIDEActionValueType ActionValueType {
		get => _actionValueType;
		set {
			if (_actionValueType == value) return;
			_actionValueType = value;
			EmitChanged();
		}
	}

	private bool _blockLowerPriorityActions = true;
	[Export]
	public bool BlockLowerPriorityActions {
		get => _blockLowerPriorityActions;
		set {
			if (_blockLowerPriorityActions == value) return;
			_blockLowerPriorityActions = value;
			EmitChanged();
		}
	}

	[ExportGroup("Godot Actions")]
	private bool _emitAsGodotActions = false;
	[Export]
	public bool EmitAsGodotActions {
		get => _emitAsGodotActions;
		set {
			if (_emitAsGodotActions == value) return;
			_emitAsGodotActions = value;
			EmitChanged();
		}
	}

	[ExportGroup("Action Remapping")]
	private bool _isRemappable;
	[Export]
	public bool IsRemappable {
		get => _isRemappable;
		set {
			if (_isRemappable == value) return;
			_isRemappable = value;
			EmitChanged();
		}
	}

	private string _displayName;
	[Export]
	public string DisplayName {
		get => _displayName;
		set {
			if (_displayName == value) return;
			_displayName = value;
			EmitChanged();
		}
	}

	private string _displayCategory;
	[Export]
	public string DisplayCategory {
		get => _displayCategory;
		set {
			if (_displayCategory == value) return;
			_displayCategory = value;
			EmitChanged();
		}
	}

	[Signal]
	public delegate void TriggeredEventHandler();

	[Signal]
	public delegate void StartedEventHandler();

	[Signal]
	public delegate void OngoingEventHandler();

	[Signal]
	public delegate void CompletedEventHandler();

	[Signal]
	public delegate void CancelledEventHandler();

	public GUIDEActionState LastState { get; internal set; } = GUIDEActionState.Completed;

	public bool ValueBool { get; private set; }
	public float ValueAxis1D => Value.X;
	public Vector2 ValueAxis2D { get; private set; } = Vector2.Zero;
	public Vector3 Value { get; private set; } = Vector3.Zero;

	public float ElapsedSeconds { get; private set; }
	public float ElapsedRatio { get; private set; }
	public float TriggeredSeconds { get; private set; }

	internal float _triggerHoldThreshold = -1.0f;

	internal void _Triggered(Vector3 value, float delta) {
		TriggeredSeconds += delta;
		ElapsedRatio = 1.0f;
		_UpdateValue(value);
		LastState = GUIDEActionState.Triggered;
		EmitSignal(SignalName.Triggered);
		_EmitGodotActionMaybe(true);
	}

	internal void _Started(Vector3 value) {
		ElapsedRatio = 0.0f;
		_UpdateValue(value);
		LastState = GUIDEActionState.Ongoing;
		EmitSignal(SignalName.Started);
		EmitSignal(SignalName.Ongoing);
	}

	internal void _Ongoing(Vector3 value, float delta) {
		ElapsedSeconds += delta;
		if (_triggerHoldThreshold > 0) {
			ElapsedRatio = ElapsedSeconds / _triggerHoldThreshold;
		}
		_UpdateValue(value);
		bool wasTriggered = LastState == GUIDEActionState.Triggered;
		LastState = GUIDEActionState.Ongoing;
		EmitSignal(SignalName.Ongoing);
		if (wasTriggered) {
			_EmitGodotActionMaybe(false);
		}
	}

	internal void _Cancelled(Vector3 value) {
		ElapsedSeconds = 0;
		ElapsedRatio = 0;
		_UpdateValue(value);
		LastState = GUIDEActionState.Completed;
		EmitSignal(SignalName.Cancelled);
		EmitSignal(SignalName.Completed);
	}

	internal void _Completed(Vector3 value) {
		ElapsedSeconds = 0;
		ElapsedRatio = 0;
		TriggeredSeconds = 0;
		_UpdateValue(value);
		LastState = GUIDEActionState.Completed;
		EmitSignal(SignalName.Completed);
		_EmitGodotActionMaybe(false);
	}

	private void _EmitGodotActionMaybe(bool pressed) {
		if (!EmitAsGodotActions) return;

		if (string.IsNullOrEmpty(Name)) {
			GD.PushError("Cannot emit action into Godot's system because name is empty.");
			return;
		}

		var godotAction = new InputEventAction();
		godotAction.Action = Name;
		godotAction.Strength = Value.X;
		godotAction.Pressed = pressed;
		Input.ParseInputEvent(godotAction);
	}

	internal void _UpdateValue(Vector3 value) {
		switch (ActionValueType) {
			case GUIDEActionValueType.Bool:
			case GUIDEActionValueType.Axis1D:
				ValueBool = Math.Abs(value.X) > 0;
				ValueAxis2D = new Vector2(Math.Abs(value.X), 0);
				Value = new Vector3(value.X, 0, 0);
				break;
			case GUIDEActionValueType.Axis2D:
				ValueBool = Math.Abs(value.X) > 0;
				ValueAxis2D = new Vector2(value.X, value.Y);
				Value = new Vector3(value.X, value.Y, 0);
				break;
			case GUIDEActionValueType.Axis3D:
				ValueBool = Math.Abs(value.X) > 0;
				ValueAxis2D = new Vector2(value.X, value.Y);
				Value = value;
				break;
		}
	}

	public bool IsTriggered() => LastState == GUIDEActionState.Triggered;
	public bool IsCompleted() => LastState == GUIDEActionState.Completed;
	public bool IsOngoing() => LastState == GUIDEActionState.Ongoing;

	public string EditorName() {
		if (!string.IsNullOrEmpty(DisplayName)) return DisplayName;
		if (!string.IsNullOrEmpty(Name)) return Name;
		return ResourcePath.GetFile().Replace(".tres", "");
	}
}
